Sources/StreamChat/WebSocketClient/WebSocketClient.swift

```diff
@@ -137,23 +141,18 @@ class WebSocketClient {
         source: WebSocketConnectionState.DisconnectionSource = .userInitiated,
         completion: @escaping () -> Void
     ) {
-        connectionState = .disconnecting(source: source)
-        engineQueue.async { [engine, eventsBatcher] in
-            engine?.disconnect()
-
-            eventsBatcher.processImmediately(completion: completion)
+        switch connectionState {
+        case .initialized, .disconnected, .disconnecting:
+            connectionState = .disconnected(source: source)
+        case .connecting, .waitingForConnectionId, .connected:
+            connectionState = .disconnecting(source: source)
```

Comment:
when disconnecting, if we are already disconnected, we should not put the state to disconnecting since the engine won't do anything, and so it won't report the disconnected afterwards. For this reason, we need to instantly report as disconnected for this scenario.

Reply:
this seems tricky, previously it seems we were also processing events. Can you explain a bit more why we delete that part? 
Also, the switch itself is a bit strange to me, especially the second case. If you are disconnected, and you are connecting, why should it go disconnecting?

Reply:
I did not change the processing events part. The github diffing is not very clear, but the only thing changed here is the report of `connectionState`

> Also, the switch itself is a bit strange to me, especially the second case. If you are disconnected, and you are connecting, why should it go disconnecting?
Where does this happen? We report disconnected if we were not connected previously, so theres nothing to disconnect, that is why we instantly report that we are disconnected. If we are connecting, or waiting for connection or connected, then we need to call disconnect to the engine, so the engine will eventually report the disconnected state.

Reply:
It is quite clear actually:

When the state is `initialized`, `.disconnected`, and `disconnecting`, and the user disconnects, we instantly report `disconnected` since we don't need to disconnect the websocket engine (it is already disconnected).

When the state is `connected`, `connecting` or `waiting for connection`, we need to call the `engine.disconnect()` and the websocket delegate will report the `disconnected` state. This is why here we report `disconnecting` instead of `disconnected`.

Reply:
yeah, now after seeing the diagram and the explanations, it makes more sense, thanks.